File: memex/internal/memex/embedder.go
package memex

import (
	"log"
	"math"
	"sort"
	"strings"
)

// Embedder computes embeddings for notes in the background.
// When embedding is disabled (default), it uses BM25-based text similarity as fallback.
type Embedder struct {
	queue   chan string
	store   *Store
	enabled bool
}

// NewEmbedder creates an Embedder. If not enabled, the queue is a no-op.
func NewEmbedder(store *Store, enabled bool) *Embedder {
	e := &Embedder{
		queue:   make(chan string, 100),
		store:   store,
		enabled: enabled,
	}
	if enabled {
		go e.processLoop()
	}
	return e
}

// Enqueue adds a note ID to the embedding queue.
func (e *Embedder) Enqueue(id string) {
	if !e.enabled {
		return
	}
	select {
	case e.queue <- id:
	default:
		log.Printf("embedder: queue full, dropping note %s", id)
	}
}

// IsEnabled returns whether embedding is active.
func (e *Embedder) IsEnabled() bool {
	return e.enabled
}

func (e *Embedder) processLoop() {
	for id := range e.queue {
		if err := e.processNote(id); err != nil {
			log.Printf("embedder: failed to process note %s: %v", id, err)
		}
	}
}

func (e *Embedder) processNote(id string) error {
	note, err := e.store.Get(id)
	if err != nil {
		return err
	}

	// Use simple bag-of-words embedding as a baseline.
	// When kelindar/search is integrated, this will be replaced with
	// proper sentence-transformer embeddings (all-MiniLM-L6-v2).
	embedding := bowEmbedding(note.Content)

	return e.store.SetEmbedding(id, embedding)
}

// bowEmbedding creates a simple bag-of-words embedding.
// This is a placeholder that provides basic semantic similarity
// until the proper sentence-transformer model is integrated.
func bowEmbedding(text string) []float32 {
	tokens := tokenize(text)
	if len(tokens) == 0 {
		return make([]float32, 384)
	}

	// Hash tokens into a fixed-size vector
	vec := make([]float32, 384)
	for _, token := range tokens {
		// Simple hash-based projection
		h := hashString(token)
		for i := 0; i < 3; i++ {
			idx := (h + uint32(i)*2654435761) % 384
			if h&(1<<uint(i)) != 0 {
				vec[idx] += 1.0
			} else {
				vec[idx] -= 1.0
			}
		}
	}

	// L2 normalize
	var norm float64
	for _, v := range vec {
		norm += float64(v) * float64(v)
	}
	if norm > 0 {
		norm = math.Sqrt(norm)
		for i := range vec {
			vec[i] = float32(float64(vec[i]) / norm)
		}
	}

	return vec
}

func hashString(s string) uint32 {
	var h uint32 = 2166136261
	for i := 0; i < len(s); i++ {
		h ^= uint32(s[i])
		h *= 16777619
	}
	return h
}

// SimilarNotes finds notes similar to the given query text using embeddings.
func (e *Embedder) SimilarNotes(query string, k int) []string {
	if !e.enabled {
		return nil
	}

	queryEmb := bowEmbedding(query)
	allEmbs := e.store.AllEmbeddings()

	type scored struct {
		id    string
		score float64
	}

	var results []scored
	for id, emb := range allEmbs {
		sim := CosineSimilarity(queryEmb, emb)
		if sim > 0.1 {
			results = append(results, scored{id, sim})
		}
	}

	sort.Slice(results, func(i, j int) bool {
		return results[i].score > results[j].score
	})

	limit := k
	if len(results) < limit {
		limit = len(results)
	}

	ids := make([]string, limit)
	for i := 0; i < limit; i++ {
		ids[i] = results[i].id
	}
	return ids
}

// FindSimilarByContent finds notes with content similar to the given text.
// Uses BM25 when embeddings are disabled, cosine similarity when enabled.
func FindSimilarByContent(store *Store, text string, k int) []string {
	notes, err := store.Search(SearchParams{Query: text})
	if err != nil || len(notes) == 0 {
		return nil
	}

	limit := k
	if len(notes) < limit {
		limit = len(notes)
	}

	ids := make([]string, limit)
	for i := 0; i < limit; i++ {
		ids[i] = notes[i].ID
	}
	return ids
}

// KeywordOverlapScore computes the fraction of shared tokens between two texts.
func KeywordOverlapScore(a, b string) float64 {
	tokensA := uniqueTokens(a)
	tokensB := uniqueTokens(b)

	if len(tokensA) == 0 || len(tokensB) == 0 {
		return 0
	}

	overlap := 0
	for t := range tokensA {
		if tokensB[t] {
			overlap++
		}
	}

	// Jaccard similarity
	union := len(tokensA) + len(tokensB) - overlap
	if union == 0 {
		return 0
	}
	return float64(overlap) / float64(union)
}

// DetectProject attempts to detect the project name and relative path for a file.
func DetectProject(filePath string) Source {
	// Normalize path
	filePath = strings.TrimSpace(filePath)
	if filePath == "" {
		return Source{}
	}

	// Simple heuristic: split on common project indicators
	// In practice, the AI provides source in "project:path" format
	if idx := strings.Index(filePath, ":"); idx >= 0 {
		return Source{
			Project: filePath[:idx],
			Path:    filePath[idx+1:],
		}
	}

	return Source{Path: filePath}
}

File: memex/src/embedder.ts
import { Store } from "./store.js";
import { cosineSimilarity } from "./search.js";

// Lazy-loaded sentence-transformer pipeline
let extractor: any = null;
let extractorFailed = false;

async function getExtractor(): Promise<any> {
  if (extractor) return extractor;
  if (extractorFailed) return null;

  try {
    const { pipeline } = await import("@huggingface/transformers");
    extractor = await pipeline("feature-extraction", "Xenova/all-MiniLM-L6-v2");
    return extractor;
  } catch (err) {
    extractorFailed = true;
    console.error("embedder: failed to load model, falling back to BoW:", err);
    return null;
  }
}

async function computeEmbedding(text: string): Promise<number[]> {
  const ext = await getExtractor();
  if (ext) {
    const output = await ext(text, { pooling: "mean", normalize: true });
    return Array.from(output.data as Float32Array);
  }
  // Fallback to BoW if model unavailable
  return bowEmbedding(text);
}

export class Embedder {
  private queue: string[] = [];
  private processing = false;
  private store: Store;
  private enabled: boolean;

  constructor(store: Store, enabled: boolean) {
    this.store = store;
    this.enabled = enabled;
  }

  isEnabled(): boolean {
    return this.enabled;
  }

  enqueue(id: string): void {
    if (!this.enabled) return;
    this.queue.push(id);
    this.processNext();
  }

  private async processNext(): Promise<void> {
    if (this.processing || this.queue.length === 0) return;
    this.processing = true;

    while (this.queue.length > 0) {
      const id = this.queue.shift()!;
      try {
        await this.processNote(id);
      } catch (err) {
        console.error(`embedder: failed to process note ${id}:`, err);
      }
    }

    this.processing = false;
  }

  private async processNote(id: string): void {
    const note = this.store.get(id);
    const embedding = await computeEmbedding(note.content);
    this.store.setEmbedding(id, embedding);
  }

  async similarNotes(query: string, k: number): Promise<string[]> {
    if (!this.enabled) return [];
    const queryEmb = await computeEmbedding(query);
    const allEmbs = this.store.allEmbeddings();

    const scored: { id: string; score: number }[] = [];
    for (const [id, emb] of Object.entries(allEmbs)) {
      const sim = cosineSimilarity(queryEmb, emb);
      if (sim > 0.1) scored.push({ id, score: sim });
    }

    scored.sort((a, b) => b.score - a.score);
    return scored.slice(0, k).map((s) => s.id);
  }
}

// --- BoW fallback ---

function bowEmbedding(text: string): number[] {
  const tokens = text.toLowerCase().match(/[a-z0-9_]+/g) ?? [];
  const vec = new Float32Array(384);

  for (const token of tokens) {
    const h = fnv1a(token);
    for (let i = 0; i < 3; i++) {
      const idx = ((h + i * 2654435761) >>> 0) % 384;
      vec[idx] += (h & (1 << i)) ? 1 : -1;
    }
  }

  let norm = 0;
  for (const v of vec) norm += v * v;
  if (norm > 0) {
    norm = Math.sqrt(norm);
    for (let i = 0; i < vec.length; i++) vec[i] /= norm;
  }

  return Array.from(vec);
}

function fnv1a(s: string): number {
  let h = 2166136261;
  for (let i = 0; i < s.length; i++) {
    h ^= s.charCodeAt(i);
    h = Math.imul(h, 16777619);
  }
  return h >>> 0;
}
